### DISCUSSION TITLE
How to make the entire Tr clickable without changing the table structure?

### DISCUSSION THREAD

**Discussion by alirazmjooei** [chakra-ui/chakra-ui#4714] [Q&A] [answered]
Hello, I want to `Link` the table row, but when i link the `Tr`, the structure of the whole table changes.
Also if i put Tr in a `Link` tag, nothing changes and the structure of the table changes again and everything is misplaced.

```html
<Tbody>
   <Tr
       as={Link}
       to={somewhere}
       cursor="pointer"
        >
       <Td>example1</Td>
       <Td>example2</Td>
       <Td>example3</Td>
  </Tr>
</Tbody>
```

**Comment by payapula** [comment, marked as answer]
Hi @ElliotCyber 

Please try to create a sandbox for your issue when posting a question here, this helps others to quickly jump on the train to help you out 😄 

Having said that, I can give you two solutions, one with `Link` component and one without `Link` by having `onClick` directly on `Tr` to redirect user via `window.location`.  If you want proper semantics, prefer the one with `Link` Component.

## Solution 1 - With Link

```jsx
<Tbody>
  <Link
      href="https://chakra-ui.com"
      isExternal
      cursor="pointer"
      display="contents" // <------ Note this style
    >
        <Tr
          _hover={{
            background: "lightblue"
          }}
        >
            <Td>yards</Td>
            <Td>metres (m)</Td>
            <Td isNumeric>0.91444</Td>
        </Tr>
    </Link>
</Tbody>
```

We want `display:contents` to render <a> out of <tbody> as a pseudo-box. Read more - [Wrapping HTML table rows in a tags](https://stackoverflow.com/a/61589399)

## Solution 2 - With onClick

```jsx
<Tbody>
    <Tr
      cursor="pointer"
      _hover={{
        background: "lightblue"
      }}
      onClick={() => (window.location = "https://chakra-ui.com")}
    >
          <Td>inches</Td>
          <Td>millimetres (mm)</Td>
          <Td isNumeric>25.4</Td>
    </Tr>
</Tbody>
```

## Codesandbox 

Live usage - https://codesandbox.io/s/brave-cdn-h7fkl?file=/src/index.js

Let me know if this helps :)

**Reply by alirazmjooei** [reply]
Hi @payapula 
Thank you for your answer, I am using the React Router `Link` so in the first solution you provided, It works but with Chakra Ui `Link` tag. Besides, I have a button and ellipsis button in each row of the table, which are using React router `Link` again. In this example, the table structure changes and columns are changed. https://codesandbox.io/s/elastic-dust-c6z0z?file=/src/index.js

For the second solution, I need to pass `onClick` to each `Td` of the table because if i pass it to `Tr`, the button + ellipsis button get the same link as it has been mentioned in `onClick`. https://codesandbox.io/s/gifted-fire-glm70?file=/src/index.js

So finally, i need to somehow link `Tr` without the structure of the table being changed!

**Reply by payapula** [reply]
@ElliotCyber 

> I need to pass onClick to each Td of the table because if i pass it to Tr, the button + ellipsis button get the same link as it has been mentioned in onClick

Hm, I understand that you need a `Link` wrapper for `Tr` that points to a different route, and there is a `button` in the same row, which points to a different route.  

Will this work for you?

https://codesandbox.io/s/youthful-cookies-vzhzq?file=/src/index.js

The gist is [chakra factory](https://chakra-ui.com/docs/features/chakra-factory)

```jsx
import { chakra } from "@chakra-ui/react";
import { Link } from "react-router-dom";

const RouterChakraLink = chakra(Link); // https://chakra-ui.com/docs/features/chakra-factory

 <RouterChakraLink display="contents" to="/contact">
    <Tr>
        <Td>inches</Td>
        <Td>millimetres (mm)</Td>
        <Td isNumeric>25.4</Td>
        <Td>
          <Link to="/about">
            <Button colorScheme="green">About</Button>
          </Link>
        </Td>
    </Tr>
 </RouterChakraLink>
```
So, when you click on the table row, it would transition to `/contact` but if you click on button, it would go to `/about`.
